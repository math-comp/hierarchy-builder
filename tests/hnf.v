From HB Require Import structures.

HB.mixin Record M A := { x: nat }.
HB.structure Definition S := { X of M X}.

HB.factory Record f A := { y : nat }.
HB.builders Context T of f T.
HB.instance Definition _ := M.Build T (y + 1).
HB.end.

#[hnf] HB.instance Definition _ := f.Build nat (3 + 2).
Print Datatypes_nat__canonical__hnf_S.
Print HB_unnamed_mixin_8.

HB.instance Definition _ := f.Build bool (3 + 2).
Print Datatypes_bool__canonical__hnf_S.
Print HB_unnamed_mixin_12.

