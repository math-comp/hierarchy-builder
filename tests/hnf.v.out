Datatypes_nat__canonical__hnf_S =
{| S.sort := nat; S.class := {| S.hnf_M_mixin := HB_unnamed_mixin_11 |} |}
     : S.type
HB_unnamed_mixin_11 =
{| M.x := f.y nat HB_unnamed_factory_9 + 1 |}
     : M.axioms_ nat
Datatypes_bool__canonical__hnf_S =
{| S.sort := bool; S.class := {| S.hnf_M_mixin := hnf_f__to__hnf_M__14 |} |}
     : S.type
hnf_f__to__hnf_M__14 =
Builders_5.HB_unnamed_factory_7 bool HB_unnamed_factory_12
     : M.axioms_ bool
